### ISSUE TITLE
Get "No Elm workspace contains..." with nested projects

### ISSUE THREAD

**Issue by chadselph** [elm-tooling/elm-language-client-vscode#94]
I checked out the elm-3d-scene project from here https://github.com/ianmackenzie/elm-3d-scene and when I was inside the examples, elm-format was not happening on save.

## Expected Behavior
Elm formatting to work on a file in the examples/ directory in elm-3d-scene.

## Current Behavior
Elm format does not work on files in the examples/ sub directory.

## Possible Solution
It's probably because it finds the `elm.json` in the parent folder and then never looks for the `elm.json` in the examples folder.

## Steps to Reproduce (for bugs)
1. git clone https://github.com/ianmackenzie/elm-3d-scene.git
2. code elm-3d-scene
3. Go into examples/ (any file)
4. Put some spaces before a line and save.

**Comment by Razzeee** [comment]
Yeah, it's not supposed to work like that, you want to open the examples either as another vscode workspace or as another vscode window.

**Comment by ret394** [comment]
thanks. good to know this happens.

**Comment by Razzeee** [comment]
I guess the sensible thing here would be to document this in the FAQ https://github.com/elm-tooling/elm-language-client-vscode#faq and close this issue thereafter.

**Comment by MedeaMelana** [comment]
I ran into this today. Why does this work this way? Requiring an inner project to be opened in its own window seems arbitrary to me.

The extension seems to handle projects that are siblings just fine, so there doesn't seem to be a problem with having multiple Elm projects within the same VS window per se.

**Comment by Razzeee** [comment]
> The extension seems to handle projects that are siblings just fine, so there doesn't seem to be a problem with having multiple Elm projects within the same VS window per se.

The problem is, that your editor window is not able to choose between the elm projects, should it open in the context of the parent or in the context of the child?

**Comment by MedeaMelana** [comment]
I would say, the context of the child, because it is the more specific one. It means you can have another project inside a bigger one and have the extension use the config of the subproject. It also means that there is a chance for each config to be used (depending on the file's location), whereas if you pick the parent context, there is no file location that would use the child config.

**Comment by Razzeee** [comment]
I would like to have it the other way around, as it makes no sense to me, that I end up not finding references to the child project otherwise.

**Comment by mewhit** [comment]
And what about tests folder ? I get same error, what is the workaround, if i want features provided by elm-languag-client for all my tests ?

**Comment by Razzeee** [comment]
haven't really looked into that

**Comment by marcosh** [comment]
Runned into this issue myself. I agree that the closest `elm.json` (i.e. the child one) should be considered. Another option is to try both... if the parent one fails, then you try the child one

**Comment by ymtszw** [comment]
An unconfirmed idea, does elm-ls work with multiple Elm apps inside a root project, as long as they are NOT nested?

Example:

```
root/
|- package.json
|- package-lock.json
|- ...
|- workspace-app/
|  |- elm.json
|  |- src/Main.elm
|  `- ...
|- workspace-review/
|  |- elm.json
|  |- src/ReviewConfig.elm
|  `- ...
|- workspace-generator/
|  |- elm.json
|  |- src/GeneratorConfig.elm
|  `- ...
`- ...
```

My intuition says yes, but not tried yet.
If it works, this style can be considered as a sort-of-workaround. Assuming the root project's package.json utilizes npm workspaces for consolidated deps management.

**Comment by ymtszw** [comment]
I have a project with nested Elm apps (one root Elm app and two child Elm apps directories) and a quick test has shown this indeed works!
What I did was temporarily move out elm.json of root Elm app, reload VSCode, and open Elm files from two different child Elm apps within one single VSCode instance.
Files from both Elm apps can be correctly handled by elm-ls. Completions, format, code lens, everything works fine.

**Comment by VladimirLogachev** [comment]
https://github.com/VladimirLogachev/elm-format-problem

This repo illustrates elm formatting issues.

When the repo is opened as a workspace, formatting doesn't work properly.

It only works for one of the apps, randomly `app1` or `app2`.

Extension output:

```
[Error - 12:17:09 AM] Error: No Elm workspace contains /Users/user/projects/elm-format-problem/app2/src/Main.elm delayed processing of request
```

**Comment by razzeee** [comment]
That seems to be a different problem, as those projects are not nested. I would guess the client gets something wrong and talks to the wrong server. As you can see, that the other server updates the text files correctly.

**Comment by VladimirLogachev** [comment]
Sure!
I just considered this discussion to be the closest to my problem and posted my example here.
Should I create another issue?

**Comment by razzeee** [comment]
Yes, but I'm not sure when I can look at it.

**Comment by kantuni** [comment]
For people who came here to fix formatting, install `elm-format` globally (using `bun`, `pnpm`, etc.) and add the following to your Zed config:

```json
"languages": {
  "Elm": {
    "formatter": {
      "external": {
        "command": "elm-format",
        "arguments": ["--stdin"],
      }
    }
  }
}
```
